Mau.DylanLaMarca

DylanLaMarca opened the pull request.

DylanLaMarca committed:
Deleted Mau JNI

Deleted the Data JNI files from the Mau subdirectory

DylanLaMarca committed:
Maintenance and VMX Refactor

Refactored the VMXPointers.h file to be called VMXHandler.h, as well as implementing the changes suggested/required

DylanLaMarca committed:
Merge remote-tracking branch 'origin/Mau.DylanLaMarca' into Mau.DylanLaMarca

DylanLaMarca committed:
Moved VMX DriverStation to Mau

Moved the DriverStation directory from the base directory of allwpilib to the Mau subdirectory

DylanLaMarca committed:
Created Channel Maps

Created two json files. One contains all of the indexes, types, and capabilities of the VMX channels and the other Maps each WPI Channel to a VMX Channel

DylanLaMarca committed:
Moved VMXHandler 

Relocated the header file that handles VMX functions for the wpi to the new Translator sub-directory of Mau

DylanLaMarca committed:
Rough Shared Memory

Created a rough class called Mau_DriveData which should contain all of
the information necessary for DriverStation communication through
BagelCpp

DylanLaMarca committed:
Added rapidJson Library

Added the rapidJson library (https://github.com/Tencent/rapidjson) to
the translator subdirectory so it can be utilized to efficiently read
the Channel and Enum maps

DylanLaMarca committed:
Added VMXHandler Include

Include VMXHandler in PWM so we can translate WPI operations to VMX
operations

DylanLaMarca committed:
Update

Quick update so Caleb and I can work on DriveData/BagelCpp at the same
time

calebmhartshorn committed:
Worked on DriveData.cpp/.h files and ds_comms to integrate with wpilibs

DylanLaMarca committed:
Finalized Channel Map and PWM Implementation

Finished the channel mapping mechanism, made (most of) the mau files
buildable, and implemented the channel mapper in PWM

DylanLaMarca committed:
Began Mau Read Conversion

Reorganized the DriverStation subdirectory, changed what DriveData
holds, and how DriveData is accessed

DylanLaMarca committed:
Rough draft of Mau_DriveData and BagelDriveData

Made DriveData utilize BagelCpp fwi structs through a new header file
BagelDriveData.h and more

calebmhartshorn committed:
Pointer bug fixes

DylanLaMarca committed:
Merge branch 'Mau.DylanLaMarca' of https://github.com/kauailabs/allwpilib into Mau.DylanLaMarca

DylanLaMarca committed:
Merge branch 'Mau.DylanLaMarca' of https://github.com/kauailabs/allwpilib into Mau.DylanLaMarca

kauailabs committed:
Added trivial Single Motor Joystick Test FRC program.

kauailabs committed:
Merge branch 'Mau.DylanLaMarca' of https://github.com/kauailabs/allwpilib into Mau.DylanLaMarca

DylanLaMarca committed:
Replaced fpga_clock

Replaced "MockHooks," which managed simulated fpgs time in the HAL sim,
with MauClock, which has equivilent functions that implement the VMXTime
object

DylanLaMarca committed:
Fully Implemented MauClock

Removed includes to MockHooks.h files and refactored the HAL initializer
function for MauClock

DylanLaMarca committed:
Basic, Rough Implementation of Mau_DriveData

Replaced implementations of Toast shared memory with Mau_DriveData in
functions required for basic functionality

DylanLaMarca committed:
Merge branch 'Mau.DylanLaMarca' of https://github.com/kauailabs/allwpilib into Mau.DylanLaMarca

DylanLaMarca committed:
Integrated Mutex into Mau_DriveData

Added priority_mutex to Mau_DriveData, using one to lock all functions
which write, update, or read any data from an instance of Mau_DriveData

calebmhartshorn committed:
ds_comms periodic update called by thread. Init-ish joystick

KadenK committed:
Modified (but not functioning) CMakeLists.txt file to include hal-mau

KadenK committed:
Merge branch 'Mau.DylanLaMarca' of https://github.com/kauailabs/allwpilib into Mau.DylanLaMarca

calebmhartshorn committed:
Delete Periodic update and mutex lock server_running

calebmhartshorn committed:
Merge branch 'Mau.DylanLaMarca' of https://github.com/kauailabs/allwpilib into Mau.DylanLaMarca

DylanLaMarca committed:
Fixed VMXTime reference

Changed vmxGetTime to call just Microseconds instead of HighPortion
Microseconds

DylanLaMarca committed:
Cleaning Build Errors

Commented out a number of small oversights which caused build errors in
non-vital parts of the HAL, including the entirety of Interrupts because
it's a saucy little minxy

DylanLaMarca committed:
Added Updaters to Mau_DriveData

Added a number of functions to Mau_DriveData and the BagelDriveData
utility to update individual ControlWord booleans and
JoystickDescriptors

DylanLaMarca committed:
Merge branch 'Mau.DylanLaMarca' of https://github.com/kauailabs/allwpilib into Mau.DylanLaMarca

KadenK committed:
Continued update of CMakeLists.txt

KadenK committed:
A working CMakeLists.txt file for the libwpiHalMau.so file only (not libwpiHal.so or any java)

DylanLaMarca committed:
Connected DriverStation, Bagel, and DriveData

DriverStation now initializes all of the comms. When created it creates
the shared instance of DriveData and HAL_WaitForDSData waits for the
shared Data's signal, and HAL_InitializeDriverStation starts the
BagelCpp's ds_comms threads

DylanLaMarca committed:
Merge branch 'Mau.DylanLaMarca' of https://github.com/kauailabs/allwpilib into Mau.DylanLaMarca

KadenK committed:
Working Java and C++ CMakeLists.txt file****

****The include files are not all present, the CMakeLists.txt file is correct, but the build environment is not. Until that is updated, FRC applications will not run successfully.

KadenK committed:
Merge branch 'Mau.DylanLaMarca' of https://github.com/kauailabs/allwpilib into Mau.DylanLaMarca

DylanLaMarca committed:
Cleaned DriverStation

Deleted the Toast shared memory stuff and integrated the structure of
ds_comms into DriverStation

DylanLaMarca committed:
Re-added Interrupts

Fixed the naming error in Interrupts init and un-commented Interrupts

DylanLaMarca committed:
Changed JSON Reference to library subdirectory

Changed the const filepaths that the FileHandler uses to locate the
json-map files to /maps, which is where cmake will store them with the
library

DylanLaMarca committed:
Merge branch 'Mau.DylanLaMarca' of https://github.com/kauailabs/allwpilib into Mau.DylanLaMarca

KadenK committed:
Renamed Hal output jar

DylanLaMarca committed:
Fixed Build Errors

Added declaration of mau::sharedMemory and added undeclared variables

DylanLaMarca committed:
Readded Extensions.cpp

Uncommented all of Extensions.cpp

KadenK committed:
Successfully building C++ sample

KadenK committed:
C++ Eclipse set-up files

kauailabs committed:
Resolved Stack Memory Allocation Errors

DylanLaMarca committed:
Rough Hack Fixes

Relocated testing isDS attached to tcp Thread, Changed Mau_DriveData to
return no pointers, but have yet to build and completely test

DylanLaMarca committed:
Scaled WPI Mau to VMXpi Timing

Changed tests of time recieved from the VMXTime to expect and test in
microseconds instead of milliseconds, as well as decreasing the expected
sample rate

kauailabs committed:
Fixed MatchInfo Memory Errors and Joystick Values

kauailabs commented on hal/src/main/native/mau/PWM.cpp:
```diff
@@ -22,342 +25,380 @@ namespace hal {
-        return handle;
+    PWMGeneratorConfig vmxConfig(200 /* Frequency in Hz */);
+    mau::vmxIO->ActivateSinglechannelResource(mauChannel->getInfo(), &vmxConfig, mauChannel->vmxResHandle,
+                                              mau::vmxError);
```
Error handling needed here.

kauailabs commented on hal/src/main/native/mau/PWM.cpp:
```diff
@@ -22,342 +25,380 @@ namespace hal {
 
-    //  SimPWMData[port->channel].SetInitialized(false);
-        digitalChannelHandles->Free(pwmPortHandle, HAL_HandleEnum::PWM);
+    std::string vmxLabel = mau::enumConverter->getHandleLabel(HAL_HandleEnum::PWM);
```
I'm thinking these three lines of code will be invoked as a group often throughout code that uses VMX-pi; consider creating some sort of helper function/method to encapsulate them.

kauailabs commented on hal/src/main/native/mau/PWM.cpp:
```diff
@@ -22,342 +25,380 @@ namespace hal {
+    std::string vmxLabel = mau::enumConverter->getHandleLabel(HAL_HandleEnum::PWM);
+    Mau_Channel* mauChannel = mau::channelMap->getChannel(vmxLabel, port->channel);
+    VMXResourceHandle vmxResource = mauChannel->vmxResHandle;
+    mau::vmxIO->UnrouteChannelFromResource(mauChannel->vmxIndex, vmxResource, mau::vmxError);
```
All of this code can be replaced with a single call to vmxIO->DeallocateResource.  Internally, this code will take care of this (and much more that is not included in the above code in PWM.cpp).

kauailabs commented on hal/src/main/native/mau/PWM.cpp:
```diff
@@ -22,342 +25,380 @@ namespace hal {
+}
+
+HAL_Bool HAL_CheckPWMChannel(int32_t channel) {
+    return channel < kNumPWMChannels && channel >= 0;
```
This is just a question:  does the json channel mapping have a concept that the total number of pwm channels is variable?  I'm assuming that kNumPWMChannels == the max possible number of pwm channels that can be mapped within the json file, please let me know if that assumption is correct.

kauailabs commented on hal/src/main/native/mau/PWM.cpp:
```diff
@@ -22,342 +25,380 @@ namespace hal {
-            *status = HAL_HANDLE_ERROR;
-            return;
-        }
+    std::string vmxLabel = mau::enumConverter->getHandleLabel(HAL_HandleEnum::PWM);
```
simplify to helper function?

kauailabs commented on hal/src/main/native/mau/PWM.cpp:
```diff
@@ -22,342 +25,380 @@ namespace hal {
 
-    //  SimPWMData[port->channel].SetRawValue(value);
-    }
+    mau::vmxIO->PWMGenerator_SetDutyCycle(mauChannel->vmxResHandle, vmxIndex, value, mau::vmxError);
```
Error handling needed.

kauailabs commented on hal/src/main/native/mau/PWM.cpp:
```diff
@@ -22,342 +25,380 @@ namespace hal {
-            *status = INCOMPATIBLE_STATE;
-            return;
-        }
+    std::string vmxLabel = mau::enumConverter->getHandleLabel(HAL_HandleEnum::PWM);
```
simplify?

kauailabs commented on hal/src/main/native/mau/PWM.cpp:
```diff
@@ -22,342 +25,380 @@ namespace hal {
-        } else if (pos > 1.0) {
-            pos = 1.0;
-        }
+    mau::vmxIO->PWMGenerator_SetDutyCycle(mauChannel->vmxResHandle, vmxIndex, dutyCycle, mau::vmxError);
```
error handling

kauailabs commented on hal/src/main/native/mau/PWM.cpp:
```diff
@@ -22,342 +25,380 @@ namespace hal {
-    //  SimPWMData[port->channel].SetRawValue(0);
-    //  SimPWMData[port->channel].SetPosition(0);
-    //  SimPWMData[port->channel].SetSpeed(0);
+    if (pos < 0.0) {
```
This is something I need to carefully review.  I'm thinking that disabling should cause a PWM output of 0 (no duty cycle at all, signal should be held low at the digital output).  I'll get back to you on this part.

kauailabs commented on hal/src/main/native/mau/PWM.cpp:
```diff
@@ -22,342 +25,380 @@ namespace hal {
+    std::string vmxLabel = mau::enumConverter->getHandleLabel(HAL_HandleEnum::PWM);
+    Mau_Channel* mauChannel = mau::channelMap->getChannel(vmxLabel, port->channel);
+    VMXResourceIndex vmxIndex = mauChannel->getResourceIndex();
+    mau::vmxIO->PWMGenerator_SetDutyCycle(mauChannel->vmxResHandle, vmxIndex, 0, mau::vmxError);
```
simplify? error handling?

kauailabs commented on hal/src/main/native/mau/PWM.cpp:
```diff
@@ -22,342 +25,380 @@ namespace hal {
+ */
+void HAL_SetPWMPeriodScale(HAL_DigitalHandle pwmPortHandle, int32_t squelchMask, int32_t* status) {
+    auto port = digitalChannelHandles->Get(pwmPortHandle, HAL_HandleEnum::PWM);
+    if (port == nullptr) {
```
We will need to add handling for this case.  It translates to setting the VMXPI HAL's PWMGeneratorConfig::SetFrameOutputFilter() method.  And since this is a configuration option, it will only take effect during initial resource configuration.  So this means that if the PWM resource is currently activated, it needs to be first deactivated, then reactivated, for these changes to take place.

kauailabs commented on hal/src/main/native/mau/DriveStation/DriverComms.cpp:
```diff
@@ -0,0 +1,199 @@
+        bool search = true;
+        int joy_id = 0;
+
+        while (i < length && search) {
```
One concern about the code as it stands now is that there are a lot of mutex lock/unlocks going on as the Mau DriveData is updated, when in principle only one lock/unlock is required.  That is not a huge problem, but is a potential performance concern.

However a second and very real functional concern is the condition variable.  The condition variable is being updated many times, and should definitely only be updated at the end when all variables have been updated.  The reason this is a real concern is that it means that reader-side code will be woken up right after the first time the condition variable is notified.  Which means the reader-side code will tend to be (based upon the functioning of the thread scheduler) attempting to read the variables at the same time this function is in the middle of updating them.  To resolve this, the condition variable should definitely only be notified when all updating is complete.  So this part is important to address soon.

kauailabs commented on hal/src/main/native/mau/DriveStation/MauDriveData.cpp:
```diff
@@ -0,0 +1,229 @@
+    memLock.lock();
+    joysticks[joyNumber].joyAxes.count = axisCount;
+    for (int index = 0; index < axisCount; index++) {
+        joysticks[joyNumber].joyAxes.axes[index] = ((float)axes[index]) / 127;
```
After review; it's required within Mau (this is a good place to do it, I think) that the value of any joystick axis cannot exceed 1.0 or fall below -1.0.  Currently, this code allows an axis value of -1.01 - so this needs to be fixed.

kauailabs reviewed (changes requested):
Code review complete; there are a few issues that need to be resolved before we can merge this work to the main branch.

kauailabs committed:
Updated simjni.gradle file to support Mau

kauailabs committed:
- Updated Mau HAL PWM-related files to work correctly.
  NOTE:  This requires the latest VMXPI-HAL (v. 1.1.148), which contains a new PWM-related API function.
- Updated ChannelMap.json and PortsInternal to properly reflect all 22 VMXPi PWM Channels.
- Update PWM.cpp to correctly select correct VMXPI PWM Port Index during initialization.
- Update PWM.cpp to support "WPI Squelched" outputs, needed for Servo support.
- Update HAL.cpp to use modified VMXPi class constructor call (w/realtime flag = true)
- Update DigitalInternal.h to support wider range (finer resolution) of PWM Steps
- Update Robot.cpp test program to simultaneously test 3 motor controllers and 1 servo.

kauailabs commented:
This pull request is being closed.  Rationale:  Scott has updated the Mau.DylanLaMarca branch with code that fixes a few PWM-related issues and updates the Mau ChannelMap.json file to support all 22 of the VMX-PI PWM Channels.  Then, a new pull request with the contents of this pull request plus these new changes is being created, to replace this pull request.
